content/post/performance-metrics-histogram.md

```diff
@@ -0,0 +1,245 @@
+
+To summarize: what we are trying to achieve is the ability to tell whether 99% or more of the events processed in a given time period have a latency below 1ms, and whether 99.9% or more of the events processed in a given time period have a latency below 1s. In addition, we need to be able to efficiently recompute this every second.
+
+## Approaches
+
+## Using A Sorted Values List
+
+Using Sorted Values List, the following method is used:
+
+1. We store all data as a list of events, L. L is sorted by time.
+2. We then select a window, W, from L, such that W contains all events whose time is within the last 5 minutes  
+    `{W: e ∀ e ∊ L | e.time > now()-'5 minutes'}`
+3. To select the values in the 99% and 99.9% positions, we need to sort W. We call the sorted list W'.
+4. We then select the 99% and 99.9% positions in the sorted list W', by taking its length and multiplying it by 0.99 and 0.999 respectively, then rounding the fraction up to the nearest integer. We use those integers as the indexes for the events whose latencies are the 99%- and 99.9%-percentiles. 
+
+### Cost Analysis
+
+- We have to store every single event in L.
+    - We have to send every single event to the monitoring hub so it can store it in L.  
+        __**Space cost: O(|L|)**__
+- We have to index L by time. 
+    - We can assume events arrive in order of completion, so L is naturally sorted.  
+      __**Time cost: 0**__
+- We have to select the window W from L.  
+    __**Space cost: O(|W|)**__  
+    __**Time cost: O(log(|W|))**__  
+- We have to sort W in order of latency values.
+    - Assuming best case performance of quicksort of O(nlogn)  
+        __**Time cost: O(|W|log(|W|))**__  
+
+__**Total time cost: O(|W|log(|W|)) + O(log|W|)**__  
+__**Total space cost: O(|L|) + O(|W|)**__
+
+## Using a Histogram
+
+Using a histogram, the following method is used: 
+
+1. We aggregate the histograms from a stream of events.
+    1. Local metrics are aggregated at the Metrics Receiver within each Wallaroo worker, and a histogram is produced per step per metrics period (typically 2 seconds).
+2. Once a metrics collection period elapses, the collection for that period is sent out to an external monitoring system.
+3. The histograms for each step are stored in a list L, sorted by time.
+4. The list for each step is stored in a hashtable H, keyed by step.
+  Note that this relationship is invertible and depends on how we prefer to query things.
+5. To get the latency histogram of a time window W, we choose all the histograms for a step and perform a bin-wise addition (since the bin ranges are identical across the histograms, this is safe). The result is the bin-wise sum of the histograms in the window, which is another histogram of the same type. We call this histogram S.
```

Comment:
It would be helpful to indicate that the histograms for a time window for a step are being selected from the hashtable created in the previous step so people are less likely to get lost.

Also, similar comment as earlier regarding the subscript to denote that `S` is for a particular step.